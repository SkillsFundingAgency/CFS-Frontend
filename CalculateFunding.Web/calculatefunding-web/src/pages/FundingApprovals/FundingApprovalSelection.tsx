import React, {useState} from "react";
import {IBreadcrumbs} from "../../types/IBreadcrumbs";
import {Header} from "../../components/Header";
import {Section} from "../../types/Sections";
import {Banner} from "../../components/Banner";
import {
    getFundingPeriodsByFundingStreamIdService,
    getFundingStreamsService,
    getSpecificationsByFundingPeriodAndStreamIdService
} from "../../services/specificationService";
import {useEffectOnce} from "../../hooks/useEffectOnce";
import {FundingPeriod, FundingStream, Specification} from "../../types/viewFundingTypes";
import {
    getFundingPeriodsByFundingStreamId,
    getSpecificationsByFundingPeriodAndStreamId
} from "../../actions/SelectSpecificationActions";
import {SpecificationSummary} from "../../types/SpecificationSummary";

export function FundingApprovalSelection() {
    let breadcrumbs: IBreadcrumbs[] = [
        {
            name: "Calculate funding",
            url: "/app"
        },
        {
            name: "Funding approvals",
            url: "/app/Approvals"
        },
        {
            name: "Select specification",
            url: null
        }
    ];

    const [fundingStreams, setFundingStreams] = useState<string[]>([]);
    const [selectedFundingStream, setSelectedFundingStream] = useState<string>("");
    const [selectedFundingPeriod, setSelectedFundingPeriod] = useState<string>("");
    const [selectedSpecification, setSelectedSpecification] = useState<string>("");
    const [fundingPeriods, setFundingPeriods] = useState<FundingPeriod[]>([]);
    const [isLoading, setIsLoading] = useState({
        fundingStreams: false,
        fundingPeriods: false,
        specification: false
    });


    useEffectOnce(() => {
        getFundingStreamsService().then((result) => {
            setFundingStreams(result.data);
        })
    });

    function populateFundingPeriod(fundingStream: string) {
        if (fundingStream === "") {
            setFundingPeriods([]);
            setSelectedSpecification("");
        } else {
            getFundingPeriodsByFundingStreamIdService(fundingStream).then((result) => {
                setFundingPeriods(result.data as FundingPeriod[])
            });
        }
    }

    function populateSpecification(fundingPeriod: string) {
        if (fundingPeriod === "") {
            setSelectedSpecification("")
        } else {
            getSpecificationsByFundingPeriodAndStreamIdService(selectedFundingStream, fundingPeriod).then((result) => {
                if (result.data.length > 0) {
                    setSelectedSpecification((result.data[0] as SpecificationSummary).id)
                }
            });
        }
    }

    function changeFundingStream(e: React.ChangeEvent<HTMLSelectElement>) {
        const fundingStream = e.target.value;
        setSelectedFundingStream(fundingStream)
        populateFundingPeriod(fundingStream)
    }


    function changeFundingPeriod(e: React.ChangeEvent<HTMLSelectElement>) {
        const fundingPeriod = e.target.value;
        setSelectedFundingPeriod(fundingPeriod)
        populateSpecification(fundingPeriod)
    }

    return <div>
        <Header location={Section.Datasets}/>
        <div className="govuk-width-container">
            <Banner bannerType="Left" breadcrumbs={breadcrumbs} title="" subtitle=""/>
            <div className="govuk-main-wrapper">
                <div className="govuk-grid-row  govuk-!-margin-bottom-9">
                    <div className="govuk-grid-column-full">
                        <h1 className="govuk-heading-xl govuk-!-margin-bottom-2">Select specification</h1>
                        <span className="govuk-caption-xl">You can select the specification and funding period.</span>
                    </div>
                </div>
                <div className="govuk-form-group">
                    <label className="govuk-label" htmlFor="sort">
                        Funding stream
                    </label>
                    <select className="govuk-select" id="funding-streams" name="sort"
                            onChange={(e) => changeFundingStream(e)}>
                        <option></option>
                        {fundingStreams.map((fs, index) => <option key={index} value={fs}>{fs}</option>)}
                    </select>
                </div>
                <div className="govuk-form-group" hidden={fundingPeriods.length === 0}>
                    <label className="govuk-label" htmlFor="sort">
                        Funding period
                    </label>
                    <select className="govuk-select" id="funding-periods" name="sort" disabled={fundingPeriods.length === 0}
                            onChange={(e) => changeFundingPeriod(e)}>
                        <option value=""></option>
                        {fundingPeriods.map((fp, index) => <option key={index} value={fp.id}>{fp.name}</option>)}
                    </select>
                </div>
                <div className="govuk-form-group" hidden={selectedSpecification === ""}>
                    <label className="govuk-label" >
                        Specification
                    </label>
                    <h3 className="govuk-heading-m">{selectedSpecification}</h3>

                    <a href={`/app/approvals/FundingApprovalResults/${selectedFundingStream}/${selectedFundingPeriod}/${selectedSpecification}`}
                       className="govuk-button" data-module="govuk-button">
                        View funding
                    </a>

                </div>
            </div>
        </div>
    </div>
}