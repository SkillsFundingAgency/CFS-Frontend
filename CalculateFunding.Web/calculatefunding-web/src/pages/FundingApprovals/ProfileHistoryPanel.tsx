import React from "react";
import {useQuery} from "react-query";
import {Link} from "react-router-dom";
import {getPreviousProfileExistsForSpecificationForProviderForFundingLine} from "../../services/fundingLineDetailsService";

export interface ProfileHistoryPanelProps {
  providerId: string;
  fundingStreamId: string;
  specificationId: string;
  fundingLineCode: string;
  providerVersionId: string;
  fundingPeriodId: string;
}

function useProfileHistory(specificationId: string, providerId: string, fundingStreamId: string, fundingLineCode: string) {
  return useQuery<boolean>(
    [`profile-history-exists`, specificationId, providerId, fundingStreamId, fundingLineCode],
    getPreviousProfileExistsForSpecificationForProviderForFundingLine, {retry: 1, enabled: specificationId});
}

export function ProfileHistoryPanel({specificationId, providerId, providerVersionId, fundingStreamId,
  fundingPeriodId, fundingLineCode}: ProfileHistoryPanelProps) {
  const {isLoading, isError, data} = useProfileHistory(specificationId, providerId, fundingStreamId, fundingLineCode);

  return (
    <>
      <h3 className="govuk-heading-m">Previous profiles</h3>
      {isLoading && <div className="govuk-inset-text">Checking whether profile history exists for this funding line...</div>}
      {isError && <div className="govuk-inset-text">An error occurred whilst checking profile history. Try refreshing the page.</div>}
      {data !== undefined && !data ? <div className="govuk-inset-text">No profile history exists for this funding line.</div> : null}
      {data !== undefined && data ?
        <div className="govuk-body">
          <Link
            to={`/Approvals/ProfilingHistory/${specificationId}/${providerId}/${providerVersionId}/${fundingStreamId}/${fundingPeriodId}/${fundingLineCode}`}
            className="govuk-link">
            History of previous profiles
        </Link>
        </div>
        : null}
    </>
  )
}