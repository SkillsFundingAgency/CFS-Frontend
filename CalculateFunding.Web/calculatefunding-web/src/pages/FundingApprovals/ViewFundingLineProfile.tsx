import React, {useEffect, useState} from "react";
import {RouteComponentProps, useHistory} from "react-router";
import {Header} from "../../components/Header";
import {Section} from "../../types/Sections";
import {Breadcrumb, Breadcrumbs} from "../../components/Breadcrumbs";
import {MultipleErrorSummary} from "../../components/MultipleErrorSummary";
import {Footer} from "../../components/Footer";
import {ErrorMessage} from "../../types/ErrorMessage";
import {getFundingLinePublishedProviderDetails, applyCustomProfile} from "../../services/publishedProviderFundingLineService";
import {Link} from "react-router-dom";
import {FundingLineProfile, FundingLineProfileViewModel} from "../../types/PublishedProvider/FundingLineProfile";
import {LoadingStatus} from "../../components/LoadingStatus";
import {FormattedNumber, NumberType} from "../../components/FormattedNumber";
import {useSelector} from "react-redux";
import {FundingStreamPermissions} from "../../types/FundingStreamPermissions";
import {IStoreState} from "../../reducers/rootReducer";
import {DateFormatter} from "../../components/DateFormatter";
import {PermissionStatus} from "../../components/PermissionStatus";
import {EditableProfileTotal} from "../../components/Funding/EditableProfileTotal";
import {cloneDeep} from "lodash";
import {ApplyCustomProfileRequest, ProfilePeriodType} from "../../types/PublishedProvider/ApplyCustomProfileRequest";
import {ProfileTotal} from "../../types/FundingLineProfile";
import {ProfileHistoryPanel} from "./ProfileHistoryPanel";

export interface ViewFundingLineProfileProps {
    providerId: string;
    fundingStreamId: string;
    fundingPeriodId: string;
    specificationId: string;
    fundingLineId: string;
    specCoreProviderVersionId: string;
}

export function ViewFundingLineProfile({match}: RouteComponentProps<ViewFundingLineProfileProps>) {
    const permissions: FundingStreamPermissions[] = useSelector((state: IStoreState) => state.userState.fundingStreamPermissions);
    const fundingStreamId = match.params.fundingStreamId;
    const fundingPeriodId = match.params.fundingPeriodId;
    const specificationId = match.params.specificationId;
    const fundingLineId = match.params.fundingLineId;
    const providerId = match.params.providerId;
    const providerVersionId = match.params.specCoreProviderVersionId;

    const history = useHistory();

    const [fundingLineProfile, setFundingLineProfile] = useState<FundingLineProfile>({
        fundingLineCode: '',
        fundingLineName: '',
        ukprn: '',
        amountAlreadyPaid: 0,
        carryOverAmount: null,
        providerName: '',
        profilePatternKey: '',
        profilePatternName: '',
        profilePatternDescription: '',
        lastUpdatedUser: {id: '', name: '', },
        profileTotalAmount: 0,
        profileTotals: []
    });
    const [editedFundingLineProfile, setEditedFundingLineProfile] = useState<FundingLineProfile>({
        fundingLineCode: '',
        fundingLineName: '',
        ukprn: '',
        amountAlreadyPaid: 0,
        carryOverAmount: null,
        providerName: '',
        profilePatternKey: '',
        profilePatternName: '',
        profilePatternDescription: '',
        lastUpdatedUser: {id: '', name: '', },
        profileTotalAmount: 0,
        profileTotals: []
    });
    const [errors, setErrors] = useState<ErrorMessage[]>([]);
    const [missingPermissions, setMissingPermissions] = useState<string[]>([]);
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const [isSaving, setIsSaving] = useState<boolean>(false);
    const [hasPermission, setHasPermission] = useState<boolean>(false);
    const [isEditMode, setIsEditMode] = useState<boolean>(false);
    const [canEditCustomProfile, setCanEditCustomProfile] = useState<boolean>(false);
    const [canChangeToRuleBasedProfile, setCanChangeToRuleBasedProfile] = useState<boolean>(false);

    useEffect(() => {
        getFundingLineProfile();
    }, []);

    useEffect(() => {
        setMissingPermissions([]);
        const fundingStreamPermission = permissions.find(p => p.fundingStreamId === fundingStreamId);
        if (!fundingStreamPermission || !fundingStreamPermission.canApplyCustomProfilePattern) {
            setMissingPermissions(["apply custom profile pattern"]);
        } else {
            setHasPermission(true);
        }
    }, [permissions]);

    const isFormValid = (totalAllocationAmount: number, totalAllocationPercent: number) => {
        let isErrors = false;
        if (totalAllocationAmount > remainingAmount) {
            addErrorMessage("Total allocation greater than balance available", "totalAllocation");
            isErrors = true;
        }
        if (totalAllocationPercent > 100) {
            addErrorMessage("Total must be less than or equal to 100", "totalPercent");
            isErrors = true;
        }
        return !isErrors;
    }

    const getFundingLineProfile = async () => {
        try {
            setIsLoading(true);
            const response = await getFundingLinePublishedProviderDetails(specificationId, providerId, fundingStreamId, fundingLineId, fundingPeriodId);
            const profile = response.data as FundingLineProfileViewModel;
            setEditedFundingLineProfile(profile.fundingLineProfile);
            setFundingLineProfile(profile.fundingLineProfile);
            setCanEditCustomProfile(profile.enableUserEditableCustomProfiles);
            setCanChangeToRuleBasedProfile(profile.enableUserEditableRuleBasedProfiles);
        }
        catch (err) {
            addErrorMessage(err.message);
        }
        finally {
            setIsLoading(false);
        }
    }

    const handleEditProfileClick = async () => {
        if (!isEditMode) {
            setIsEditMode(true);
            window.scrollTo(0, 0);
        } else {
            try {
                clearErrorMessages(["totalPercent", "totalAllocation"]);
                const totalUnpaidAllocationAmount = calculateTotalUnpaidAllocationAmount();
                const totalUnpaidAllocationPercent = calculateUnpaidTotalAllocationPercent();

                if (!isFormValid(totalUnpaidAllocationAmount, totalUnpaidAllocationPercent)) {
                    return;
                }

                setIsSaving(true);

                const carryForwardValue = calculateNewCarryForwardAmount(totalUnpaidAllocationAmount);
                const request: ApplyCustomProfileRequest = {
                    fundingStreamId: fundingStreamId,
                    fundingPeriodId: fundingPeriodId,
                    fundingLineCode: fundingLineId,
                    providerId: providerId,
                    customProfileName: `${providerId}-${fundingStreamId}-${fundingPeriodId}-${fundingLineId}`,
                    carryOver: carryForwardValue > 0 ? carryForwardValue - originalCarryForwardAmount : null,
                    profilePeriods: editedFundingLineProfile ? editedFundingLineProfile.profileTotals.map(pt => ({
                        type: pt.periodType as ProfilePeriodType,
                        typeValue: pt.typeValue,
                        year: pt.year,
                        occurrence: pt.occurrence,
                        profiledValue: pt.value,
                        distributionPeriodId: pt.distributionPeriodId
                    })) : []
                };

                await applyCustomProfile(request);
                await getFundingLineProfile();
                setIsEditMode(false);
            } catch (err) {
                addErrorMessage(err.message);
            } finally {
                setIsSaving(false);
            }
        }
    }

    const handleCancelClick = () => {
        setEditedFundingLineProfile(fundingLineProfile);
        setIsEditMode(false);
    }

    const handleChangeToRuleBasedProfileClick = () => {
        history.push(`/Approvals/ProviderFundingOverview/${specificationId}/${providerId}/${providerVersionId}/${fundingStreamId}/${fundingPeriodId}/${fundingLineId}/change-profile-type`);
    }

    function updateProfileTotal(instalmentNumber: number, newProfileTotal: ProfileTotal) {
        if (!editedFundingLineProfile) return;
        const cloneOfFundingLineProfile: FundingLineProfile = cloneDeep(editedFundingLineProfile);
        cloneOfFundingLineProfile.profileTotals = cloneOfFundingLineProfile.profileTotals.map(
            profile => profile.installmentNumber === instalmentNumber ? newProfileTotal : profile
        );
        setEditedFundingLineProfile(cloneOfFundingLineProfile);
    }

    function addErrorMessage(errorMessage: string, fieldName?: string) {
        const errorCount: number = errors.length;
        const error: ErrorMessage = {id: errorCount + 1, fieldName: fieldName, message: errorMessage};
        setErrors(errors => [...errors, error]);
    }

    function clearErrorMessages(fieldNames?: string[]) {
        if (errors.length > 0) {
            if (fieldNames === undefined) {
                setErrors([]);
            } else {
                setErrors(errors.filter(e => !e.fieldName || (e.fieldName && !fieldNames.includes(e.fieldName))));
            }
        }
    }

    function calculateUnpaidTotalAllocationPercent(): number {
        if (!editedFundingLineProfile || editedFundingLineProfile.profileTotals.length === 0) return 0;
        const totalPercentage = editedFundingLineProfile.profileTotals
            .filter(p => p.profileRemainingPercentage !== undefined && !p.isPaid)
            .map(p => p.profileRemainingPercentage)
            .reduce((a, c) => (a !== undefined && c !== undefined ? a + c : 0), 0);

        return totalPercentage || 0;
    }

    function calculateTotalUnpaidAllocationAmount(): number {
        if (!editedFundingLineProfile || editedFundingLineProfile.profileTotals.length === 0) return 0;
        const totalAllocation = editedFundingLineProfile.profileTotals
            .filter(p => !p.isPaid)
            .map(p => p.value)
            .reduce((a, c) => a + c, 0);

        return totalAllocation || 0;
    }

    function calculateTotalPaidAndUnpaidAllocationAmount(): number {
        if (!editedFundingLineProfile || editedFundingLineProfile.profileTotals.length === 0) return 0;
        const totalAllocation = editedFundingLineProfile.profileTotals
            .map(p => p.value)
            .reduce((a, c) => a + c, 0);

        return totalAllocation || 0;
    }

    function calculateNewCarryForwardAmount(totalUnpaidAllocationAmount: number): number {
        return remainingAmount - totalUnpaidAllocationAmount;
    }

    const fundingLineName = fundingLineProfile && fundingLineProfile.fundingLineName ?
        fundingLineProfile.fundingLineName : "Missing funding line name";
    const remainingAmount = fundingLineProfile &&
        fundingLineProfile.remainingAmount !== undefined && fundingLineProfile.remainingAmount !== null ?
        fundingLineProfile.remainingAmount : 0;
    const originalCarryForwardAmount = fundingLineProfile && fundingLineProfile.carryOverAmount !== null ?
        fundingLineProfile.carryOverAmount : 0;
    const totalUnpaidAllocationAmount = calculateTotalUnpaidAllocationAmount();
    const totalUnpaidAllocationPercent = calculateUnpaidTotalAllocationPercent();
    const totalAllocationAmount = calculateTotalPaidAndUnpaidAllocationAmount();
    const newCarryForwardAmount = calculateNewCarryForwardAmount(totalUnpaidAllocationAmount);

    return (
        <div>
            <Header location={Section.Approvals} />
            <div className="govuk-width-container">
                {isLoading || isSaving ?
                    <LoadingStatus title={`${isLoading ? "Loading" : "Saving"} funding line profile`} /> :
                    <>
                        <Breadcrumbs>
                            <Breadcrumb name="Calculate funding" url={"/"} />
                            <Breadcrumb name="Approvals" />
                            <Breadcrumb name="Select specification" url={"/Approvals/Select"} />
                            <Breadcrumb name={"Funding approval results"} url={`/Approvals/SpecificationFundingApproval/${fundingStreamId}/${fundingPeriodId}/${specificationId}`} />
                            <Breadcrumb name={fundingLineProfile.providerName} url={`/Approvals/ProviderFundingOverview/${specificationId}/${providerId}/${providerVersionId}/${fundingStreamId}/${fundingPeriodId}`} />
                            <Breadcrumb name={fundingLineName} />
                        </Breadcrumbs>
                        <div className="govuk-grid-row">
                            <div className="govuk-grid-column-full">
                                <PermissionStatus requiredPermissions={missingPermissions} hidden={isLoading} />
                            </div>
                        </div>
                        <div className="govuk-grid-row">
                            <div className="govuk-grid-column-full">
                                <MultipleErrorSummary errors={errors} />
                            </div>
                        </div>
                        <div>
                            <div className="govuk-grid-row govuk-!-margin-bottom-5 govuk-!-margin-top-5">
                                <div className="govuk-grid-column-two-thirds">
                                    <h1 className="govuk-heading-xl govuk-!-margin-bottom-2" data-testid="funding-line-name">{fundingLineName}</h1>
                                    <h3 className="govuk-heading-m govuk-!-margin-bottom-2" data-testid="provider-name">{fundingLineProfile.providerName}</h3>
                                    <p className="govuk-body-s"><span data-testid="ukprn">{`UKPRN: ${fundingLineProfile.ukprn}`}</span></p>
                                    <br />
                                    <p className="govuk-body-s" data-testid="last-updated-by">
                                        {`Last updated by ${fundingLineProfile.lastUpdatedUser.name} on `}
                                        {fundingLineProfile.lastUpdatedDate && <DateFormatter date={fundingLineProfile.lastUpdatedDate} utc={false} />}
                                    </p>
                                </div>
                            </div>
                            <div className="govuk-grid-row">
                                <div className="govuk-grid-column-two-thirds">
                                    <span className="govuk-caption-m">Total allocation</span>
                                    <h2 className="govuk-heading-m govuk-!-margin-bottom-2" data-testid="total-allocation">
                                        <FormattedNumber value={fundingLineProfile.totalAllocation || 0} type={NumberType.FormattedMoney} />
                                    </h2>
                                    <span className="govuk-caption-m">Instalments paid value</span>
                                    <h2 className="govuk-heading-m govuk-!-margin-bottom-2" data-testid="amount-already-paid">
                                        <FormattedNumber value={fundingLineProfile.amountAlreadyPaid} type={NumberType.FormattedMoney} />
                                    </h2>
                                    <span className="govuk-caption-m">Balance available for profiling</span>
                                    <h2 className="govuk-heading-m govuk-!-margin-bottom-2" data-testid="remaining-amount">
                                        <FormattedNumber value={remainingAmount} type={NumberType.FormattedMoney} />
                                    </h2>
                                    <span className="govuk-caption-m">Balance to be carried forward</span>
                                    <h2 className="govuk-heading-m govuk-!-margin-bottom-2" data-testid="balance-carried-forward">
                                        <FormattedNumber value={originalCarryForwardAmount} type={NumberType.FormattedMoney} />
                                    </h2>
                                </div>
                            </div>
                            <div className="govuk-grid-row">
                                <div className="govuk-grid-column-full">
                                    <table className="govuk-table govuk-!-margin-top-5" data-testid="data-table">
                                        <caption className="govuk-table__caption">Profiling instalments</caption>
                                        <thead className="govuk-table__head">
                                            <tr className="govuk-table__row">
                                                <th scope="col" className="govuk-table__header">Instalment</th>
                                                <th scope="col" className="govuk-table__header">Payment status</th>
                                                <th scope="col" className="govuk-table__header">Instalment number</th>
                                                <th scope="col" className="govuk-table__header">Per cent</th>
                                                <th scope="col" className="govuk-table__header govuk-table__header--numeric">Value</th>
                                            </tr>
                                        </thead>
                                        <tbody className="govuk-table__body">
                                            {editedFundingLineProfile.profileTotals
                                                .sort((a, b) => a.installmentNumber - b.installmentNumber)
                                                .map((p, i) => (
                                                    <tr className="govuk-table__row" key={p.installmentNumber} data-testid="profile-total">
                                                        <th scope="row" className="govuk-table__header">
                                                            {p.actualDate && <DateFormatter date={p.actualDate} />}
                                                        </th>
                                                        <td className="govuk-table__cell" data-testid={`paid-${i}`}>{p.isPaid ? <strong className="govuk-tag">Paid</strong> : null}</td>
                                                        <td className="govuk-table__cell" data-testid={`instalment-number-${i}`}>{p.installmentNumber}</td>
                                                        <EditableProfileTotal
                                                            index={i}
                                                            profileTotal={p}
                                                            remainingAmount={editedFundingLineProfile.remainingAmount || 0}
                                                            setProfileTotal={updateProfileTotal}
                                                            isEditMode={isEditMode}
                                                            errors={errors}
                                                            addErrorMessage={addErrorMessage}
                                                            clearErrorMessages={clearErrorMessages} />
                                                    </tr>))}
                                            <tr className="govuk-table__row">
                                                <th scope="row" className="govuk-table__header">
                                                    Total allocation
                                                </th>
                                                <td className="govuk-table__cell"></td>
                                                <td className="govuk-table__cell"></td>
                                                <td className="govuk-table__cell" id="totalPercent">
                                                    <div className={`govuk-form-group ${errors.filter(e =>
                                                        e.fieldName === "totalPercent").length > 0 ? 'govuk-form-group--error' : ''}`}>
                                                        {errors.filter(e => e.fieldName === "totalPercent").length > 0 ?
                                                            errors.map(error => error.fieldName === "totalPercent" &&
                                                                <span key={error.id} className="govuk-error-message govuk-!-margin-bottom-1">
                                                                    <span className="govuk-visually-hidden">Error:</span> {error.message}
                                                                </span>
                                                            ) : null}
                                                        <strong data-testid="total-allocation-percent">
                                                            <FormattedNumber value={totalUnpaidAllocationPercent} type={NumberType.FormattedPercentage}
                                                                decimalPlaces={totalUnpaidAllocationPercent === 100 ? 0 : 2} />
                                                        </strong>
                                                    </div>
                                                </td>
                                                <td className="govuk-table__cell govuk-table__cell--numeric" id="totalAllocation">
                                                    <div className={`govuk-form-group ${errors.filter(e =>
                                                        e.fieldName === "totalAllocation").length > 0 ? 'govuk-form-group--error' : ''}`}>
                                                        {errors.filter(e => e.fieldName === "totalAllocation").length > 0 ?
                                                            errors.map(error => error.fieldName === "totalAllocation" &&
                                                                <span key={error.id} className="govuk-error-message govuk-!-margin-bottom-1">
                                                                    <span className="govuk-visually-hidden">Error:</span> {error.message}
                                                                </span>
                                                            ) : null}
                                                        <strong data-testid="total-allocation-amount">
                                                            <FormattedNumber value={totalAllocationAmount} type={NumberType.FormattedMoney} />
                                                        </strong>
                                                    </div>
                                                </td>
                                            </tr>
                                            <tr className="govuk-table__row">
                                                <th scope="row" className="govuk-table__header">
                                                    To be carried forward
                                                </th>
                                                <td className="govuk-table__cell"></td>
                                                <td className="govuk-table__cell"></td>
                                                <td className="govuk-table__cell"></td>
                                                <td className="govuk-table__cell govuk-table__cell--numeric">
                                                    <strong data-testid="balance-carried-forward-2">
                                                        <FormattedNumber value={newCarryForwardAmount} type={NumberType.FormattedMoney} />
                                                    </strong>
                                                </td>
                                            </tr>
                                        </tbody>
                                    </table>
                                </div>
                            </div>
                            <div className="govuk-grid-row">
                                <div className="govuk-grid-column-two-thirds">
                                    {canEditCustomProfile && <button className="govuk-button govuk-!-margin-right-1" disabled={!hasPermission}
                                        onClick={handleEditProfileClick} data-testid="edit-profile-btn">
                                        {isEditMode ? "Apply profile" : "Edit profile"}
                                    </button>}
                                    {isEditMode && <button className="govuk-button govuk-button--secondary govuk-!-margin-right-1" onClick={handleCancelClick} data-testid="cancel-btn">
                                        Cancel
                                    </button>}
                                    {canChangeToRuleBasedProfile &&
                                        <button className="govuk-button" onClick={handleChangeToRuleBasedProfileClick}>Change to rule based profile</button>}
                                </div>
                            </div>
                        </div>
                        <hr className="govuk-section-break govuk-section-break--m govuk-section-break--visible" />
                        <div className="govuk-grid-row">
                            <div className="govuk-grid-column-two-thirds">
                                <ProfileHistoryPanel
                                    specificationId={specificationId} providerId={providerId} providerVersionId={providerVersionId}
                                    fundingStreamId={fundingStreamId} fundingPeriodId={fundingPeriodId} fundingLineCode={fundingLineId} />
                                <Link to={`/Approvals/ProviderFundingOverview/${specificationId}/${providerId}/${providerVersionId}/${fundingStreamId}/${fundingPeriodId}`}
                                    className="govuk-back-link">
                                    Back
                                </Link>
                            </div>
                        </div>
                    </>}
            </div>
            <Footer />
        </div>
    );
}