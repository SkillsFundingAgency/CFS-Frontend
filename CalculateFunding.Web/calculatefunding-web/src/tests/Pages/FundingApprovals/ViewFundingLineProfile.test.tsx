import React from "react";
import * as redux from "react-redux";
import {match, MemoryRouter} from "react-router";
import {createLocation, createMemoryHistory} from "history";
import {ViewFundingLineProfileProps} from "../../../pages/FundingApprovals/ViewFundingLineProfile";
import {getAllByTestId, render, waitFor} from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';

const useSelectorSpy = jest.spyOn(redux, 'useSelector');

describe("<ViewFundingLineProfile />", () => {
    beforeAll(() => {
        jest.mock("../../../services/publishedProviderFundingLineService", () => mockPublishedProviderFundingLineService());
    });

    describe("when user has canEditProfilePattern permission", () => {
        beforeAll(() => {
            useSelectorSpy.mockReturnValue([{
                fundingStreamId: "fundingStreamId",
                canEditProfilePattern: true
            }]);
        });

        afterAll(() => {
            useSelectorSpy.mockClear();
        });

        it("does not show a permissions message", async () => {
            const {container} = renderPage();
            await waitFor(() => {
                expect(container.querySelector("#permission-alert-message")).not.toBeInTheDocument();
            });
        });

        it("shows the correct funding line profile values", async () => {
            const {getByTestId, getAllByTestId} = renderPage();
            await waitFor(() => {
                expect(getByTestId('funding-line-name')).toHaveTextContent("name")
                expect(getByTestId("provider-name")).toHaveTextContent("provider");
                expect(getByTestId("ukprn")).toHaveTextContent("UKPRN: 12345");
                expect(getByTestId("last-updated-by")).toHaveTextContent("Last updated by test user on 10 September 1907 0:00 am");
                expect(getByTestId("total-allocation")).toHaveTextContent("£100.00");
                expect(getByTestId("amount-already-paid")).toHaveTextContent("£50.00");
                expect(getByTestId("remaining-amount")).toHaveTextContent("£50.00");
                expect(getByTestId("balance-carried-forward")).toHaveTextContent("£50.00");
                expect(getAllByTestId("profile-total")).toHaveLength(2);
                expect(getAllByTestId("profile-total")).toMatchSnapshot();
            });
        });
    });

    describe("when user does not have canEditProfilePattern permission", () => {
        beforeAll(() => {
            useSelectorSpy.mockReturnValue([{
                fundingStreamId: "fundingStreamId",
                canEditProfilePattern: false
            }]);
        });

        afterAll(() => {
            useSelectorSpy.mockClear();
        });

        it("does not show a permissions message", async () => {
            const {container} = renderPage();
            await waitFor(() => {
                expect(container.querySelector("#permission-alert-message")).not.toBeInTheDocument();
            });
        });
    })
});

const renderPage = () => {
    const {ViewFundingLineProfile} = require("../../../pages/FundingApprovals/ViewFundingLineProfile");
    return render(
        <MemoryRouter>
            <ViewFundingLineProfile match={matchMock} location={location} history={history} />
        </MemoryRouter>
    );
}

const history = createMemoryHistory();
const location = createLocation("", "", "");
const matchMock: match<ViewFundingLineProfileProps> = {
    params: {
        specificationId: "specId",
        providerId: "providerId",
        providerVersionId: "providerVersionId",
        fundingLineId: "fundingLineId",
        fundingPeriodId: "fundingPeriodId",
        fundingStreamId: "fundingStreamId"
    },
    path: "",
    isExact: true,
    url: ""
};

function mockPublishedProviderFundingLineService() {
    const originalService = jest.requireActual("../../../services/publishedProviderFundingLineService");
    return {
        ...originalService,
        getFundingLinePublishedProviderDetails: jest.fn(() => Promise.resolve({
            data: {
                fundingLineCode: "123",
                fundingLineName: "name",
                ukprn: "12345",
                totalAllocation: 100,
                amountAlreadyPaid: 50,
                remainingAmount: 50,
                carryOverAmount: 50,
                providerName: "provider",
                profilePatternKey: "key",
                profilePatternName: "pattern",
                profilePatternDescription: "description",
                lastUpdatedUser: {
                    id: "1",
                    name: "test user"
                },
                lastUpdatedDate: new Date(2, 2, 2020),
                profileTotalAmount: 150,
                profileTotals: [{
                    year: 2020,
                    typeValue: "April",
                    occurrence: 1,
                    value: 50,
                    periodType: "CalendarMonth",
                    isPaid: true,
                    installmentNumber: 1,
                    profileRemainingPercentage: 24.99,
                    actualDate: new Date(1, 1, 2020)
                },
                {
                    year: 2020,
                    typeValue: "April",
                    occurrence: 2,
                    value: 50,
                    periodType: "CalendarMonth",
                    isPaid: false,
                    installmentNumber: 2,
                    profileRemainingPercentage: 24.99,
                    actualDate: new Date(2, 1, 2020)
                }]
            }
        }))
    }
}
