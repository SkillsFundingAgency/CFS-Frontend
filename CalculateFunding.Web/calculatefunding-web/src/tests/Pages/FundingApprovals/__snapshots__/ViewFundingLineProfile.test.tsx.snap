// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewFundingLineProfile /> when user has canEditProfilePattern permission shows the correct funding line profile values 1`] = `
Array [
  <tr
    class="govuk-table__row"
    data-testid="profile-total"
  >
    <th
      class="govuk-table__header"
      scope="row"
    >
      <span>
        13
         
        August
         
        1906
      </span>
    </th>
    <td
      class="govuk-table__cell"
    >
      <strong
        class="govuk-tag"
      >
        Paid
      </strong>
    </td>
    <td
      class="govuk-table__cell"
    >
      1
    </td>
    <td
      class="govuk-table__cell"
    >
      <span>
        25%
      </span>
    </td>
    <td
      class="govuk-table__cell govuk-table__cell--numeric"
    >
      <span>
        £50.00
      </span>
    </td>
  </tr>,
  <tr
    class="govuk-table__row"
    data-testid="profile-total"
  >
    <th
      class="govuk-table__header"
      scope="row"
    >
      <span>
        13
         
        August
         
        1907
      </span>
    </th>
    <td
      class="govuk-table__cell"
    />
    <td
      class="govuk-table__cell"
    >
      2
    </td>
    <td
      class="govuk-table__cell"
    >
      <span>
        25%
      </span>
    </td>
    <td
      class="govuk-table__cell govuk-table__cell--numeric"
    >
      <span>
        £50.00
      </span>
    </td>
  </tr>,
]
`;
